### PR TITLE
[NTOS] Fix / assert proper locking for VADs

### PR DESCRIPTION
**Pull request by tkreuzer**
## Purpose

Fix locking for VAD tables. Fixes CORE-18519.
The problem was identified by @Doug-Lyons. This PR extends his original fix to make sure locking is correct and asserted in all cases.

JIRA issue: [CORE-18519](https://jira.reactos.org/browse/CORE-18519)

## Proposed changes
- Add ASSERTS to MmLockAddressSpace to guarantee lock ordering
- Fix address space locking in MiProtectVirtualMemory
- Handle page table faults in MmArmAccessFault to avoid recursive aquisition of working-set lock
- Fix bugs in MmAccessFault (proper locking and fixed range check)
- Revert previous x64 hack (Temporarily release AddressCreationLock in MmCreateVirtualMappingUnsafeEx)
- Fix MmFreeMemoryArea (proper process attach, acquire working-set lock)
- Attach to the target process in MmMapViewOfSection
- Lock kernel address space in MiInitSystemMemoryAreas
- Add ASSERTs for VAD table locking

### CODE REVIEW

**Comment by binarymaster** [comment]
This PR is going to supersede #5658 and #5735, did I got it right?

File: ntoskrnl/include/internal/mm.h [hunk 1]
```diff
@@ -1690,6 +1690,12 @@ FORCEINLINE
 VOID
 MmLockAddressSpace(PMMSUPPORT AddressSpace)
 {
+    ASSERT(!PsGetCurrentThread()->OwnsProcessWorkingSetExclusive);
+    ASSERT(!PsGetCurrentThread()->OwnsProcessWorkingSetShared);
+    ASSERT(!PsGetCurrentThread()->OwnsSystemWorkingSetExclusive);
+    ASSERT(!PsGetCurrentThread()->OwnsSystemWorkingSetShared);
+    ASSERT(!PsGetCurrentThread()->OwnsSessionWorkingSetExclusive);
+    ASSERT(!PsGetCurrentThread()->OwnsSessionWorkingSetShared);
```
Comment (SergeGautherie) [inline ntoskrnl/include/internal/mm.h:1698]: Would it be worth to use a local variable for `PsGetCurrentThread()`?

File: ntoskrnl/mm/ARM3/vadnode.c [hunk 2]
```diff
@@ -43,6 +43,73 @@ CHAR MmReadWrite[32] =
 
 /* FUNCTIONS ******************************************************************/
 
+extern MM_AVL_TABLE MiRosKernelVadRoot;;
```
Comment (SergeGautherie) [inline ntoskrnl/mm/ARM3/vadnode.c:46]:
```suggestion
extern MM_AVL_TABLE MiRosKernelVadRoot;
```

File: ntoskrnl/mm/ARM3/vadnode.c [hunk 3]
```diff
@@ -43,6 +43,73 @@ CHAR MmReadWrite[32] =
...
+        ASSERT(KeGetCurrentThread()->ApcState.Process == &PsGetCurrentProcess()->Pcb);
+        ASSERT(PsGetCurrentThread()->OwnsProcessWorkingSetExclusive);
+        ASSERT(PsGetCurrentProcess()->AddressCreationLock.Owner == KeGetCurrentThread());
+    }
+}
+
+#define ASSERT_LOCKED_FOR_READ(Table) MiDbgAssertIsLockedForRead(Table)
+#define ASSERT_LOCKED_FOR_WRITE(Table) MiDbgAssertIsLockedForWrite(Table)
+
+#else
```
Comment (SergeGautherie) [inline ntoskrnl/mm/ARM3/vadnode.c:106]:
```suggestion
#else // DBG
```

File: ntoskrnl/mm/ARM3/vadnode.c [hunk 4]
```diff
@@ -43,6 +43,73 @@ CHAR MmReadWrite[32] =
...
+
+#define ASSERT_LOCKED_FOR_READ(Table)
+#define ASSERT_LOCKED_FOR_WRITE(Table)
+
+#endif
```
Comment (SergeGautherie) [inline ntoskrnl/mm/ARM3/vadnode.c:111]:
```suggestion
#endif // DBG
```

File: ntoskrnl/mm/ARM3/vadnode.c [hunk 5]
```diff
@@ -43,6 +43,73 @@ CHAR MmReadWrite[32] =
 
 /* FUNCTIONS ******************************************************************/
 
+extern MM_AVL_TABLE MiRosKernelVadRoot;;
+
+#if DBG
+static
+VOID
+MiDbgAssertIsLockedForRead(PMM_AVL_TABLE Table)
```
Comment (SergeGautherie) [inline ntoskrnl/mm/ARM3/vadnode.c:51]:
```suggestion
MiDbgAssertIsLockedForRead(_In_ PMM_AVL_TABLE Table)
```

File: ntoskrnl/mm/ARM3/vadnode.c [hunk 6]
```diff
@@ -43,6 +43,73 @@ CHAR MmReadWrite[32] =
...
+           the current process' AddressCreationLock */
+        ASSERT(Table == &PsGetCurrentProcess()->VadRoot);
+        ASSERT(MI_WS_OWNER(PsGetCurrentProcess()) ||
+               (PsGetCurrentProcess()->AddressCreationLock.Owner == KeGetCurrentThread()));
+    }
+}
+
+static
+VOID
+MiDbgAssertIsLockedForWrite(PMM_AVL_TABLE Table)
```
Comment (SergeGautherie) [inline ntoskrnl/mm/ARM3/vadnode.c:78]:
```suggestion
MiDbgAssertIsLockedForWrite(_In_ PMM_AVL_TABLE Table)
```

File: ntoskrnl/mm/marea.c [hunk 7]
```diff
@@ -357,14 +351,24 @@ MmFreeMemoryArea(
...
         }
         else
         {
+            MiLockWorkingSet(PsGetCurrentThread(), &MmSystemCacheWs);
             MiRemoveNode((PMMADDRESS_NODE)&MemoryArea->VadNode, &MiRosKernelVadRoot);
+            MiUnlockWorkingSet(PsGetCurrentThread(), &MmSystemCacheWs);
+        }
+
+        if (Process != NULL &&
+            Process != CurrentProcess)
```
Comment (SergeGautherie) [inline ntoskrnl/mm/marea.c:369]:
```suggestion
        if (Process != NULL && Process != CurrentProcess)
```

File: ntoskrnl/mm/mmfault.c [hunk 8]
```diff
@@ -235,22 +235,44 @@ MmAccessFault(IN ULONG FaultCode,
...
         }
     }
 
     /* Is this an ARM3 memory area, or is there no address space yet? */
-    if (((MemoryArea) && (MemoryArea->Type == MEMORY_AREA_OWNED_BY_ARM3)) ||
-            (!(MemoryArea) && ((ULONG_PTR)Address >= (ULONG_PTR)MmPagedPoolStart)) ||
-            (!MmGetKernelAddressSpace()))
+    if (IsArm3Fault ||
+        ((MemoryArea == NULL) && ((ULONG_PTR)Address >= (ULONG_PTR)MmPagedPoolStart)
+            && ((ULONG_PTR)Address < (ULONG_PTR)MmPagedPoolEnd)) ||
```
Comment (SergeGautherie) [inline ntoskrnl/mm/mmfault.c:274]:
```suggestion
        ((MemoryArea == NULL) &&
         ((ULONG_PTR)Address >= (ULONG_PTR)MmPagedPoolStart) &&
         ((ULONG_PTR)Address < (ULONG_PTR)MmPagedPoolEnd)) ||
```

File: ntoskrnl/include/internal/mm.h (hunk 1 above)
Comment (tkreuzer) [reply ntoskrnl/include/internal/mm.h:1698]: no

**Pull request merged**